### PR TITLE
CloudWatch: Move SessionCache onto the instance

### LINKED ISSUE #82194: CloudWatch: Audit/Update for multitenancy
**What is the chore?**:
Check the CloudWatch code (and its dependencies) to make sure they're secure for multitenancy. Remember to check over the imports from `aws-sdk-go` and how we're using it.

```[tasklist]
### Tasks
- [ ] https://github.com/grafana/grafana/pull/83278
- [ ] https://github.com/grafana/grafana/pull/83332
- [ ] https://github.com/grafana/grafana/pull/83333
- [ ] https://github.com/grafana/grafana/pull/83335
```

Notes from initial audit:
- `QueriesTotalCounter` was approved as a global ([slack link](https://raintank-corp.slack.com/archives/C05NHPPPPNE/p1707237005718529))

**Is there anything else we need to know?**:

For reference, the key parts of supporting multi-tenancy are:
- State and sensitive information should only be stored at the instance level
- Code should not reference environment variables
- Code should not access the file system

### PR DESCRIPTION
**Pull request by iwysiu**
**What is this feature?**

Updates CloudWatch to support multitenancy by storing SessionCache on the instance, removing the unused RegionCache, and removing the environment variable reference.

**Which issue(s) does this PR fix?**:

Part of #82194 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/tsdb/cloudwatch/get_dimension_values_for_wildcards.go [hunk 1]
```diff
@@ -14,13 +14,8 @@ import (
...
 // getDimensionValues gets the actual dimension values for dimensions with a wildcard
 func (e *cloudWatchExecutor) getDimensionValuesForWildcards(ctx context.Context, pluginCtx backend.PluginContext, region string,
-	client models.CloudWatchMetricsAPIProvider, origQueries []*models.CloudWatchQuery, tagValueCache *cache.Cache) ([]*models.CloudWatchQuery, error) {
-	instance, err := e.getInstance(ctx, pluginCtx)
-	if err != nil {
-		return nil, err
-	}
-
-	metricsClient := clients.NewMetricsClient(client, instance.Settings.GrafanaSettings.ListMetricsPageLimit)
+	client models.CloudWatchMetricsAPIProvider, origQueries []*models.CloudWatchQuery, tagValueCache *cache.Cache, listMetricsPageLimit int) ([]*models.CloudWatchQuery, error) {
```
Comment (iwysiu) [inline pkg/tsdb/cloudwatch/get_dimension_values_for_wildcards.go:17]: The function that calls this already calls getInstance, so I refactored it to pass in the page limit instead.

File: pkg/tsdb/cloudwatch/cloudwatch.go [hunk 2]
```diff
@@ -322,14 +315,23 @@ func (e *cloudWatchExecutor) newSession(ctx context.Context, pluginCtx backend.P
...
-		err = proxy.New(instance.ProxyOpts).ConfigureSecureSocksHTTPProxy(sess.Config.HTTPClient.Transport.(*http.Transport))
+		err = proxy.New(ds.ProxyOpts).ConfigureSecureSocksHTTPProxy(sess.Config.HTTPClient.Transport.(*http.Transport))
 		if err != nil {
 			return nil, fmt.Errorf("error configuring Secure Socks proxy for Transport: %w", err)
 		}
 	}
 	return sess, nil
 }
 
+func (e *cloudWatchExecutor) newSessionFromContext(ctx context.Context, pluginCtx backend.PluginContext, region string) (*session.Session, error) {
```
[the remark this answers is not in the document]
Comment (sarahzinger) [reply pkg/tsdb/cloudwatch/cloudwatch.go:324]: Why would it be contexts? Sorry I don't think I follow!

File: pkg/tsdb/cloudwatch/routes/external_id.go [hunk 3]
```diff
@@ -17,8 +15,13 @@ type ExternalIdResponse struct {
 }
 
 func ExternalIdHandler(ctx context.Context, pluginCtx backend.PluginContext, reqCtxFactory models.RequestContextFactoryFunc, parameters url.Values) ([]byte, *models.HttpError) {
+	reqCtx, err := reqCtxFactory(ctx, pluginCtx, "default")
+	if err != nil {
+		return nil, models.NewHttpError("error in NamespacesHandler", http.StatusInternalServerError, err)
+	}
+
 	response := ExternalIdResponse{
-		ExternalId: os.Getenv(awsds.GrafanaAssumeRoleExternalIdKeyName),
```
Comment (sarahzinger) [inline pkg/tsdb/cloudwatch/routes/external_id.go:21]: I feel like we could break up some of this code into smaller prs maybe to de-risk it a bit? For example it seems like maybe this bit could be it's own pr? Just an idea

**Review by sarahzinger — APPROVED** [review]
Nice work! ⭐ 🌟 ✨ I did a spot check and tried running the code and it seemed ok to me! It does make me a bit nervous to have us changing so much at once, so if you see ways of breaking this up into smaller prs it might be nice? But also it seems ok to me to ship as is.

File: pkg/tsdb/cloudwatch/routes/external_id.go (hunk 3 above)
Comment (iwysiu) [reply pkg/tsdb/cloudwatch/routes/external_id.go:21]: Done. Split the externalID bit into https://github.com/grafana/grafana/pull/83332 and the regionCache bit into https://github.com/grafana/grafana/pull/83333

File: pkg/tsdb/cloudwatch/cloudwatch.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (iwysiu) [reply pkg/tsdb/cloudwatch/cloudwatch.go:324]: I was thinking Contexts because it gets the instance from `ctx` and `pluginCtx` but this isn't an exported function so the naming doesn't matter that much as long as we get it.

**Review by iwysiu — COMMENTED** [review]
I'll do a commit after this to remove the stuff that was split off

File: pkg/tsdb/cloudwatch/cloudwatch.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sarahzinger) [reply pkg/tsdb/cloudwatch/cloudwatch.go:326]: ah I see, I guess I have no preference!
